### PR TITLE
Improve subheadings assessment

### LINKED ISSUE #1303: Without subheadings no warning that you should add a subheading every 300 words
If you have a subheading in your text and you write more than 300 words below that subheading, you get a red bullet point. The warning in this case is that you should use a subheading at least once every 300 words. But if you're not using any subheadings, there's no warning that you should use headings if your have more than 300 words. 

* Platform: WordPress 
* Platform version: 4.9
* Yoast SEO version: 5.9 rc1

### PR DESCRIPTION
**Pull request by nataliashitova**
## Summary
This PR can be summarized in the following changelog entry:

* Subheadings length assessment now returns feedback in all cases: for short texts <300 words the assessment always returns good scores; for longer texts the assessment checks for subheadings, if there are no subheadings the assessment returns a bad score; if there are subheadings and the text is split by them in sections of less than 300 words, the assessment returns a good score; if there are subheadings, but the text is split by them in sections of more than 300 words, the assessment returns a red score (if the longest fragment is >350 words) or an orange score (if the longest fragment is between 300 and 350 words).
* The assessment now also checks for the length of introductory text sections (before the first subheading) and reports on them as on the subheading texts.

## Relevant technical choices:

* Feedback returned to the user was adjusted to the new rules.

## Test instructions

This PR can be tested by following these steps:

### Browserified

- check out the branch
- use the browserfied example. Don't forget to run a grunt build:js
- set locale to any language or leave default
- type in a short text (<300 words) without subheadings - the assessment should return a green bullet "You are not using any subheadings, but your text is short enough and probably doesn't need them."
- add a subheading in the text (e.g., type in `<h2> Some subheading </h2>`) - the assessment should return a green bullet "Great job with using subheadings!"
- try adding a subheading in different places in your text (in the very beginning, in the very end, in the middle part) - the assessment result should not change
- clear the text field
- type in a long text (>300 words) without subheadings - the assessment should return a red bullet "You are not using any subheadings, although your text is rather long. Try and add some subheadings."
- add a subheading in the end of the text and type in a short section of text below the subheading - the assessment should return a red or orange bullet "1 section of your text is longer than 300 words and is not separated by any subheadings. Add subheadings to improve readability." The color of the bullet will depend on whether the long section of your text is below 350 words (then you get orange) or above 350 words (then you get red).
- increase the length of your post-subheading text (make it above 300 words) - the feedback should change to "2 sections of your text are longer than 300 words and are not separated by any subheadings. Add subheadings to improve readability."
- shorten both pre-subheading and post-subheading sections of your text so that each of them is about 200 words (>300 words in total) - the assessment should return a green bullet "Great job with using subheadings!"

### Wordpress

- Run yarn install in the plugin's directory of wordpress-seo.
- Run yarn add "git+https://github.com/Yoast/yoastseo.js#stories/1303-fix-subheadings-bug" in the plugin's directory.
- Navigate to node_modules/yoastseo and run yarn install and grunt build:js.
- Navigate back to the plugin's directory and run grunt build:js.
- Test the same functionality as described in the test instructions for Browserified.
- Alternatively, you can use the beta script to test the same combination of branches.

Fixes #1303

### CODE REVIEW

File: src/assessments/readability/subheadingDistributionTooLongAssessment.js [hunk 1]
```diff
@@ -23,10 +25,50 @@ class SubheadingsDistributionTooLong extends Assessment {
...
+				// The maximum recommended value of the subheading text.
+				recommendedMaximumWordCount: 300,
+				slightlyTooMany: 300,
+				farTooMany: 350,
+			},
+			url: "<a href='https://yoa.st/headings' target='_blank'>",
+			goodShortTextNoSubheadings: {
+				score: 9,
+				resultText: "You are not using any %1$ssubheadings%2$s, but your text is short enough and probably doesn't need them.",
+				resultTextPlural: "",
```
Comment (manuelaugustin) [inline src/assessments/readability/subheadingDistributionTooLongAssessment.js]: Any reason for setting the empty `resultTextPlural` strings? I think they can just be removed when they're not required.

File: src/assessments/readability/subheadingDistributionTooLongAssessment.js [hunk 2]
```diff
@@ -23,10 +25,50 @@ class SubheadingsDistributionTooLong extends Assessment {
...
+			parameters: {
+				// The maximum recommended value of the subheading text.
+				recommendedMaximumWordCount: 300,
+				slightlyTooMany: 300,
+				farTooMany: 350,
+			},
+			url: "<a href='https://yoa.st/headings' target='_blank'>",
+			goodShortTextNoSubheadings: {
+				score: 9,
+				resultText: "You are not using any %1$ssubheadings%2$s, but your text is short enough and probably doesn't need them.",
```
Comment (manuelaugustin) [inline src/assessments/readability/subheadingDistributionTooLongAssessment.js]: Very nice new feedback strings! 🙂

File: src/assessments/readability/subheadingDistributionTooLongAssessment.js [hunk 3]
```diff
@@ -38,27 +80,38 @@ class SubheadingsDistributionTooLong extends Assessment {
...
 	 * @returns {AssessmentResult} The assessment result.
 	 */
 	getResult( paper, researcher, i18n ) {
-		let subheadingTextsLength = researcher.getResearch( "getSubheadingTextLengths" );
-		subheadingTextsLength = subheadingTextsLength.sort(
-			function( a, b ) {
-				return b.wordCount - a.wordCount;
-			}
-		);
+		this._subheadingTextsLength = researcher.getResearch( "getSubheadingTextLengths" );
```
Comment (manuelaugustin) [inline src/assessments/readability/subheadingDistributionTooLongAssessment.js]: `const subheadingTextsLength = researcher.getResearch( "getSubheadingTextLengths" );`

File: src/assessments/readability/subheadingDistributionTooLongAssessment.js [hunk 4]
```diff
@@ -38,27 +80,38 @@ class SubheadingsDistributionTooLong extends Assessment {
...
 
-		let tooLongTexts = this.getTooLongSubheadingTexts( subheadingTextsLength ).length;
-		let score = this.calculateScore( subheadingTextsLength );
+		this._subheadingTextsLength = this._subheadingTextsLength.sort( function( a, b ) {
```
Comment (manuelaugustin) [inline src/assessments/readability/subheadingDistributionTooLongAssessment.js]: `this._subheadingTextsLength = subheadingTextsLength .sort` etc.

File: spec/assessments/subheadingDistributionTooLongSpec.js [hunk 5]
```diff
@@ -4,64 +4,136 @@ import Factory from "../helpers/factory.js";
 import Mark from "../../js/values/Mark.js";
 let i18n = Factory.buildJed();
 
-let paper = new Paper();
 let subheadingDistributionTooLong = new SubheadingDistributionTooLong();
 
-describe( "An assessment for scoring too long text fragments following a subheading.", function() {
-	it( "scores 3 text fragments, 0 are too long", function() {
-		let assessment = subheadingDistributionTooLong.getResult( paper, Factory.buildMockResearcher( [ { text: "", wordCount: 60 },  { text: "", wordCount: 100 }, { text: "", wordCount: 300 } ] ), i18n );
+const shortText = "a ".repeat( 200 );
```
Comment (manuelaugustin) [inline spec/assessments/subheadingDistributionTooLongSpec.js]: That's a pretty neat trick! 😄

**Comment by manuelaugustin** [comment]
CR 🚧

* `subheadingDistributionTooLongAssessment.js` doesn't have full test coverage
* @atimmer Could you have a quick look at the proposed changes and give us some feedback on whether this amount of refactoring is okay for a minor version change? For example the format  of the config values has changed. Could that potentially lead to any problems or is that acceptable?
* See comments for some other feedback points.

File: src/assessments/readability/subheadingDistributionTooLongAssessment.js [hunk 6]
```diff
@@ -38,27 +80,38 @@ class SubheadingsDistributionTooLong extends Assessment {
...
+			return b.wordCount - a.wordCount;
+		} );
+
+		this._tooLongTexts = this.getTooLongSubheadingTexts();
+		this._tooLongTextsNumber = this.getTooLongSubheadingTexts().length;
 
 		let assessmentResult = new AssessmentResult();
+		assessmentResult.setIdentifier( this.identifier );
+
+		if ( this.isApplicable( paper ) ) {
```
[the remark this answers is not in the document]
Comment (atimmer) [reply src/assessments/readability/subheadingDistributionTooLongAssessment.js]: You should never need to call `isApplicable` in the actual execution of the assessment. If that is not working, then that should be fixed.

**Comment by atimmer** [comment]
@manuelaugustin Doing a big refactor in general is not a problem for a minor version. But if there are backward incompatible changes then you'd want to a do a major version release.

I would say the constructors of assessments are not public API yet. Because they are not listed in https://github.com/Yoast/YoastSEO.js/blob/develop/index.js. But a good call to think about this! I would want the assessment to be public API at some point so flagging it if we want to change their API is a good idea.

File: src/assessments/readability/subheadingDistributionTooLongAssessment.js (hunk 1 above)
Comment (nataliashitova) [reply src/assessments/readability/subheadingDistributionTooLongAssessment.js]: Fixed.

File: src/stringProcessing/getIntroParagraphOutsideSubheadings.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * Matching this in a regex is pretty hard, since we need to find a way for matching the text after a heading, and before the end of the text.
+ * The hard thing capturing this is with a capture, it captures the next subheading as well, so it skips the next part of the text,
+ * since the subheading is already matched.
+ * For now we use this method to be sure we capture the right blocks of text. We remove all | 's from text,
+ * then replace all headings with a | and split on a |.
+ */
+	text = text.replace( /\|/ig, "" );
+	text = text.replace( /<h([1-6])(?:[^>]+)?>(.*?)<\/h\1>/ig, "|" );
+	const subheadings =  text.split( "|" );
```
[the remark this answers is not in the document]
Comment (nataliashitova) [reply src/stringProcessing/getIntroParagraphOutsideSubheadings.js]: Agreed.

File: src/assessments/readability/subheadingDistributionTooLongAssessment.js (hunk 3 above)
Comment (nataliashitova) [reply src/assessments/readability/subheadingDistributionTooLongAssessment.js]: I like to keep the research results as an attribute of the class, this way I can reach them from other functions within this class.

File: src/assessments/readability/subheadingDistributionTooLongAssessment.js (hunk 4 above)
Comment (nataliashitova) [reply src/assessments/readability/subheadingDistributionTooLongAssessment.js]: see previous comment

File: src/assessments/readability/subheadingDistributionTooLongAssessment.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (nataliashitova) [reply src/assessments/readability/subheadingDistributionTooLongAssessment.js]: Fixed the unnecessary isApplicable call inside the assessment.

**Comment by manuelaugustin** [comment]
CR done 👍

**Comment by manuelaugustin** [comment]
Moved to blocked until we can confirm whether the current class refactoring works with Jed/translator comments

**Comment by manuelaugustin** [comment]
Requires re-refactoring analogously to the SEO assessments.

**Comment by manuelaugustin** [comment]
CR & acceptance done 👍

**Pull request merged**